README.md
# Database controller

[![CII Best Practices](https://bestpractices.coreinfrastructure.org/projects/5643/badge)](https://bestpractices.coreinfrastructure.org/projects/5643)
[![e2e](https://github.com/DoodleScheduling/k8sdb-controller/workflows/e2e/badge.svg)](https://github.com/DoodleScheduling/k8sdb-controller/actions)
[![report](https://goreportcard.com/badge/github.com/DoodleScheduling/k8sdb-controller)](https://goreportcard.com/report/github.com/DoodleScheduling/k8sdb-controller)
[![license](https://img.shields.io/github/license/DoodleScheduling/k8sdb-controller.svg)](https://github.com/DoodleScheduling/k8sdb-controller/blob/master/LICENSE)
[![release](https://img.shields.io/github/release/DoodleScheduling/k8sdb-controller/all.svg)](https://github.com/DoodleScheduling/k8sdb-controller/releases)

Kubernetes Controller for database and user provisioning.
Currently the controller supports Postgres and MongoDB (as well as MongoDB Atlas).
Using the controller you can deploy databases and users defined as code on top of kubernetes.
How to deploy database servers is out of scope of this project.

## Example for PostgreSQL

Example of how to deploy a Postgres database called my-app as well as a user to the server localhost:5432.

```yaml
apiVersion: v1
kind: Secret
metadata:
  name: postgresql-admin-credentials
  namespace: default
data:
  password: MTIzNA==
  username: MTIzNA==
---
apiVersion: dbprovisioning.infra.doodle.com/v1beta1
kind: PostgreSQLDatabase
metadata:
  name: my-app
  namespace: default
spec:
  address: "postgres://localhost:5432"
  rootSecret:
    name: postgresql-admin-credentials
---
apiVersion: dbprovisioning.infra.doodle.com/v1beta1
kind: PostgreSQLUser
metadata:
  name: my-app
  namespace: default
spec:
  database:
    name: my-app
  credentials:
    name: my-app-postgresql-credentials
---
apiVersion: v1
kind: Secret
metadata:
  name: my-app-postgresql-credentials
  namespace: default
data:
  password: MTIzNA==
  username: MTIzNA==
```

## Example for MongoDB

Example of how to deploy a MongoDB database called my-app as well as a user to the server localhost:5432.

```yaml
apiVersion: v1
kind: Secret
metadata:
  name: mongodb-admin-credentials
  namespace: default
data:
  password: MTIzNA==
  username: MTIzNA==
---
apiVersion: dbprovisioning.infra.doodle.com/v1beta1
kind: MongoDBDatabase
metadata:
  name: my-app
  namespace: default
spec:
  address: "mongodb://localhost:27017"
  rootSecret:
    name: mongodb-admin-credentials
---
apiVersion: dbprovisioning.infra.doodle.com/v1beta1
kind: MongoDBUser
metadata:
  name: my-app
  namespace: default
spec:
  database:
    name: my-app-mongodb-credentials
  credentials:
    name: my-app-mongodb
  roles:
    - name: readWrite
---
apiVersion: v1
kind: Secret
metadata:
  name: my-app-mongodb-credentials
  namespace: default
data:
  password: MTIzNA==
  username: MTIzNA==
```

## Installation

### Helm

Please see [chart/k8sdb-controller](https://github.com/DoodleScheduling/k8sdb-controller/tree/master/chart/k8stcpmap-controller) for the helm chart docs.

### Manifests/kustomize

Alternatively you may get the bundled manifests in each release to deploy it using kustomize or use them directly.

## Limitations

By design there is no garbage collection implemented for databases. Meaning a database does not get dropped if the kubernetes resources is removed.
However this is not the case for users. Users will be removed from the corresponding databases if the referenced kubernetes resource gets removed.
We might reconsider this in the future.

## Profiling
To profile controller, access web server on #profilerPort (default 6060).

In Kubernetes, port-forward to this port, and open the `/debug/pprof` URL in browser. For example, if you port-forward 6060 from container to 6060 on your machine, access:
```
http://localhost:6060/debug/pprof/
```

## Configure the controller

You may change base settings for the controller using env variables (or alternatively command line arguments).
Available env variables:

| Name  | Description | Default |
|-------|-------------| --------|
| `METRICS_ADDR` | The address of the metric endpoint binds to. | `:9556` |
| `PROBE_ADDR` | The address of the probe endpoints binds to. | `:9557` |
| `ENABLE_LEADER_ELECTION` | Enable leader election for controller manager. | `false` |
| `LEADER_ELECTION_NAMESPACE` | Change the leader election namespace. This is by default the same where the controller is deployed. | `` |
| `NAMESPACES` | The controller listens by default for all namespaces. This may be limited to a comma delimited list of dedicated namespaces. | `` |
| `CONCURRENT` | The number of concurrent reconcile workers.  | `1` |


CHANGELOG.md
# Changelog

## 0.1.0

**Release date:** 2022-02-24

Initial oss release


main.go
/*


Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package main

import (
	"flag"
	"net/http"
	_ "net/http/pprof"

	"github.com/spf13/pflag"
	"github.com/spf13/viper"

	"os"
	"strings"

	"sigs.k8s.io/controller-runtime/pkg/cache"
	"sigs.k8s.io/controller-runtime/pkg/healthz"

	"k8s.io/apimachinery/pkg/runtime"
	clientgoscheme "k8s.io/client-go/kubernetes/scheme"
	_ "k8s.io/client-go/plugin/pkg/client/auth/gcp"
	ctrl "sigs.k8s.io/controller-runtime"
	"sigs.k8s.io/controller-runtime/pkg/log/zap"

	infrav1beta1 "github.com/doodlescheduling/k8sdb-controller/api/v1beta1"
	"github.com/doodlescheduling/k8sdb-controller/controllers"
	// +kubebuilder:scaffold:imports
)

var (
	scheme   = runtime.NewScheme()
	setupLog = ctrl.Log.WithName("setup")
)

// flags
const (
	MetricAddr              = "metrics-addr"
	ProbeAddr               = "probe-addr"
	ProfilerAddr            = "profiler-addr"
	EnableLeaderElection    = "enable-leader-election"
	LeaderElectionNamespace = "leader-election-namespace"
	Namespaces              = "namespaces"
	MaxConcurrentReconciles = "max-concurrent-reconciles"
)

// config variables & defaults
var (
	metricsAddr             = ":9556"
	probesAddr              = ":9557"
	profilerAddr            = ":6060"
	enableLeaderElection    = false
	leaderElectionNamespace = ""
	namespacesConfig        = ""
	maxConcurrentReconciles = 1
)

func init() {
	_ = clientgoscheme.AddToScheme(scheme)

	_ = infrav1beta1.AddToScheme(scheme)
	// +kubebuilder:scaffold:scheme
}

func main() {
	flag.StringVar(&metricsAddr, MetricAddr, metricsAddr, "The address the metric endpoint binds to.")
	flag.StringVar(&probesAddr, ProbeAddr, probesAddr, "The address of the probe endpoints bind to.")
	flag.StringVar(&profilerAddr, ProfilerAddr, profilerAddr, "The address of the profiler endpoints bind to.")
	flag.BoolVar(&enableLeaderElection, EnableLeaderElection, enableLeaderElection,
		"Enable leader election for controller manager. "+
			"Enabling this will ensure there is only one active controller manager.")
	flag.StringVar(&leaderElectionNamespace, LeaderElectionNamespace, leaderElectionNamespace, "Leader election namespace. Default is the same namespace as controller.")
	flag.StringVar(&namespacesConfig, Namespaces, namespacesConfig, "Comma-separated list of namespaces to watch. If not set, all namespaces will be watched.")
	flag.IntVar(&maxConcurrentReconciles, MaxConcurrentReconciles, maxConcurrentReconciles, "Maximum number of concurrent reconciles. Default is 1.")
	pflag.CommandLine.AddGoFlagSet(flag.CommandLine)
	pflag.Parse()

	ctrl.SetLogger(zap.New(zap.UseDevMode(true)))

	err := viper.BindPFlags(pflag.CommandLine)
	if err != nil {
		setupLog.Error(err, "Failed parsing command line arguments")
		os.Exit(1)
	}
	replacer := strings.NewReplacer("-", "_")
	viper.SetEnvKeyReplacer(replacer)
	viper.AutomaticEnv()

	metricsAddr = viper.GetString(MetricAddr)
	enableLeaderElection = viper.GetBool(EnableLeaderElection)
	leaderElectionNamespace = viper.GetString(LeaderElectionNamespace)
	probesAddr = viper.GetString(ProbeAddr)
	namespacesConfig = viper.GetString(Namespaces)
	maxConcurrentReconciles = viper.GetInt(MaxConcurrentReconciles)
	profilerAddr = viper.GetString(ProfilerAddr)

	namespaces := strings.Split(namespacesConfig, ",")
	options := ctrl.Options{
		Scheme:                 scheme,
		MetricsBindAddress:     metricsAddr,
		HealthProbeBindAddress: probesAddr,
		Port:                   9443,
		LeaderElection:         enableLeaderElection,
		LeaderElectionID:       "99a96989.doodle.com",
	}
	if len(namespaces) > 0 && namespaces[0] != "" {
		options.NewCache = cache.MultiNamespacedCacheBuilder(namespaces)
		setupLog.Info("watching configured namespaces", "namespaces", namespaces)
	} else {
		setupLog.Info("watching all namespaces")
	}
	if leaderElectionNamespace != "" {
		options.LeaderElectionNamespace = leaderElectionNamespace
	}

	// Profiler
	go func() {
		setupLog.Info("Starting profiler...")
		if err := http.ListenAndServe(profilerAddr, nil); err != nil {
			setupLog.Error(err, "Profiler failed to start")
		}
	}()

	mgr, err := ctrl.NewManager(ctrl.GetConfigOrDie(), options)
	if err != nil {
		setupLog.Error(err, "unable to start manager")
		os.Exit(1)
	}

	// Liveness probe
	err = mgr.AddHealthzCheck("healthz", healthz.Ping)
	if err != nil {
		os.Exit(1)
	}

	// Readiness probe
	err = mgr.AddReadyzCheck("readyz", healthz.Ping)
	if err != nil {
		os.Exit(1)
	}

	// MongoDBDatabase setup
	if err = (&controllers.MongoDBDatabaseReconciler{
		Client:   mgr.GetClient(),
		Log:      ctrl.Log.WithName("controllers").WithName("MongoDBDatabase"),
		Scheme:   mgr.GetScheme(),
		Recorder: mgr.GetEventRecorderFor("MongoDBDatabase"),
	}).SetupWithManager(mgr, maxConcurrentReconciles); err != nil {
		setupLog.Error(err, "unable to create controller", "controller", "MongoDBDatabase")
		os.Exit(1)
	}

	// MongoDBUser setup
	if err = (&controllers.MongoDBUserReconciler{
		Client:   mgr.GetClient(),
		Log:      ctrl.Log.WithName("controllers").WithName("MongoDBUser"),
		Scheme:   mgr.GetScheme(),
		Recorder: mgr.GetEventRecorderFor("MongoDBUser"),
	}).SetupWithManager(mgr, maxConcurrentReconciles); err != nil {
		setupLog.Error(err, "unable to create controller", "controller", "MongoDBUser")
		os.Exit(1)
	}

	// PostgreSQLDatabase setup
	if err = (&controllers.PostgreSQLDatabaseReconciler{
		Client:   mgr.GetClient(),
		Log:      ctrl.Log.WithName("controllers").WithName("PostgreSQLDatabase"),
		Scheme:   mgr.GetScheme(),
		Recorder: mgr.GetEventRecorderFor("PostgreSQLDatabase"),
	}).SetupWithManager(mgr, maxConcurrentReconciles); err != nil {
		setupLog.Error(err, "unable to create controller", "controller", "PostgreSQLDatabase")
		os.Exit(1)
	}

	// PostgreSQLUser setup
	if err = (&controllers.PostgreSQLUserReconciler{
		Client:   mgr.GetClient(),
		Log:      ctrl.Log.WithName("controllers").WithName("PostgreSQLUser"),
		Scheme:   mgr.GetScheme(),
		Recorder: mgr.GetEventRecorderFor("PostgreSQLUser"),
	}).SetupWithManager(mgr, maxConcurrentReconciles); err != nil {
		setupLog.Error(err, "unable to create controller", "controller", "PostgreSQLUser")
		os.Exit(1)
	}

	// +kubebuilder:scaffold:builder

	setupLog.Info("starting manager")
	if err := mgr.Start(ctrl.SetupSignalHandler()); err != nil {
		setupLog.Error(err, "problem running manager")
		os.Exit(1)
	}
}
